File: src/main/java/com/ryderbelserion/discord/api/commands/interfaces/CommandActor.java
package com.ryderbelserion.discord.api.commands.interfaces;

import com.ryderbelserion.discord.api.commands.CommandContext;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.SelfUser;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.jetbrains.annotations.NotNull;

public interface CommandActor {

    /**
     * Sends a simple string with an ephemeral toggle.
     */
    void reply(@NotNull final String message, final boolean ephemeral);

    /**
     * Sends an embed with an ephemeral toggle.
     */
    void reply(@NotNull final MessageEmbed embed, final boolean ephemeral);

    /**
     * Defer a reply, Sends the thinking... message to a user.
     *
     * @return the object
     */
    CommandContext defer(final boolean ephemeral);

    /**
     * Gets the option from the event.
     */
    OptionMapping getOption(@NotNull final String option);

    /**
     * @author of the message.
     */
    User getAuthor();

    /**
     * @return the one who created the bot.
     */
    User getCreator();

    /**
     * Checks if the user is the creator of the bot.
     * @param id the id
     * @return true or false
     */
    boolean isCreator(final String id);

    /**
     * @return the bot.
     */
    SelfUser getBot();

    /**
     * @return guild the command is executed in.
     */
    Guild getGuild();

    /**
     * @return jda instance.
     */
    JDA getJDA();

}

File: src/main/kotlin/com/ryderbelserion/common/command/interfaces/CommandActor.kt
package com.ryderbelserion.common.command.interfaces

import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.SelfUser
import net.dv8tion.jda.api.entities.User

abstract class CommandActor {

    abstract fun reply(message: String, ephemeral: Boolean)

    abstract fun reply(embed: MessageEmbed, ephemeral: Boolean = false)

    abstract fun author(): User?

    abstract fun creator(): User?

    abstract fun bot(): SelfUser

    abstract fun guild(): Guild?

    abstract fun jda(): JDA

}